File: scripts/scenarios/scenarios/20_unspecced_search.py
"""Scenario 20: "Beyond the Spec" — Unspecced Search & Discovery

Marcus searches across actors, posts, and starter packs using the
unspecced skeleton search endpoints.

Services: PDS, AppView (optional — some endpoints skip if no AppView)
"""

from __future__ import annotations

import sys
import time
from pathlib import Path

_project_root = str(Path(__file__).resolve().parent.parent.parent.parent)
if _project_root not in sys.path:
    sys.path.insert(0, _project_root)

from scripts.lib.atproto import XrpcClient, get_character, PDS1, ScenarioResult, timed_call



def _now() -> str:
    return time.strftime("%Y-%m-%dT%H:%M:%SZ", time.gmtime())


def run() -> ScenarioResult:
    result = ScenarioResult("Unspecced Search & Discovery")
    result.start()

    client = XrpcClient(PDS1)

    timed_call(result, "Server health check",
               lambda: client.wait_for_healthy(timeout=30))
    if result.failed > 0:
        result.finish()
        return result

    # ── Create accounts ──────────────────────────────────────────────
    char_names = ["luna", "marcus", "rosa"]
    for name in char_names:
        char = get_character(name)
        session = timed_call(
            result, f"Create account: {char.name}",
            lambda c=char: client.accounts.create_account(c.handle, c.email, c.password),
            detail_fn=lambda s, n=name: f"did={s['did']}",
        )
        if session:
            char.did = session["did"]
            char.access_jwt = session["accessJwt"]

    active = [n for n in char_names if get_character(n).did]
    if len(active) < 2:
        result.step_failed("Account creation", "Not enough accounts")
        result.finish()
        return result

    # ── Set up profiles with searchable names ────────────────────────
    for name in active:
        char = get_character(name)
        timed_call(result, f"Set profile: {char.name}",
                   lambda c=char: client.records.create_record(
                       c.did, "app.bsky.actor.profile",
                       {"$type": "app.bsky.actor.profile", "displayName": c.name, "description": c.persona},
                       c.access_jwt),
                   skip_on_status={404})

    luna = get_character("luna")
    marcus = get_character("marcus")
    rosa = get_character("rosa")

    # ── Create searchable posts ─────────────────────────────────────
    for name, texts in [
        ("luna", [
            "The Orion Nebula is absolutely stunning tonight! #astrophotography",
            "Just published my deep space photography guide #astronomy",
        ]),
        ("marcus", [
            "ATProto is the future of decentralized social networking",
            "Building a firehose consumer in Go — streaming thousands of events per second",
        ]),
        ("rosa", [
            "Homemade sourdough with roasted garlic and herbs #baking",
            "The best cacio e pepe recipe you will ever try #cooking",
        ]),
    ]:
        char = get_character(name)
        if char.did and char.access_jwt:
            for text in texts:
                timed_call(
                    result, f"{char.name} posts",
                    lambda c=char, t=text: client.records.create_record(
                        c.did, "app.bsky.feed.post",
                        {"$type": "app.bsky.feed.post", "text": t, "createdAt": _now()},
                        c.access_jwt,
                    ),
                    detail_fn=lambda r: f"uri={r['uri']}",
                )

    # ── Rosa creates a starter pack ──────────────────────────────────
    if rosa.did and rosa.access_jwt and luna.did and marcus.did:
        timed_call(
            result, "Rosa creates starter pack for search",
            lambda: client.records.create_record(
                rosa.did, "app.bsky.graph.starterpack",
                {
                    "$type": "app.bsky.graph.starterpack",
                    "name": "Space & Code Enthusiasts",
                    "description": "Friends who love space and technology",
                    "createdAt": _now(),
                },
                rosa.access_jwt,
            ),
            detail_fn=lambda r: f"uri={r['uri']}",
        )
    else:
        result.step_skipped("Rosa creates starter pack for search", "Missing DIDs")

    time.sleep(2)

    # ── Search actors skeleton ───────────────────────────────────────
    if not marcus.access_jwt:
        result.step_failed("Search", "Marcus not created")
        result.finish()
        return result

    for query in ["nebula", "Luna"]:
        timed_call(
            result, f"Search actors skeleton '{query}'",
            lambda q=query: client.search.search_actors_skeleton(q, token=marcus.access_jwt),
            detail_fn=lambda r, q=query: f"found={len(r.get('actors', []))}",
            skip_on_status={404},
        )

    # ── Search posts skeleton ────────────────────────────────────────
    for query in ["nebula", "sourdough", "zzz nonexistent content"]:
        timed_call(
            result, f"Search posts skeleton '{query}'",
            lambda q=query: client.search.search_posts_skeleton(q, token=marcus.access_jwt),
            detail_fn=lambda r, q=query: f"found={len(r.get('posts', []))}",
            skip_on_status={404},
        )

    # ── Search starter packs skeleton ────────────────────────────────
    timed_call(
        result, "Search starter packs skeleton 'Space'",
        lambda: client.search.search_starter_packs_skeleton("Space", token=marcus.access_jwt),
        detail_fn=lambda r: f"found={len(r.get('starterPacks', []))}",
        skip_on_status={404},
    )

    # ── Record artifacts ─────────────────────────────────────────────
    result.record_artifact("accounts", {
        n: {"did": get_character(n).did}
        for n in char_names if get_character(n).did
    })
    result.record_artifact("search_queries", {
        "actors_skeleton": ["nebula", "Luna"],
        "posts_skeleton": ["nebula", "sourdough", "zzz nonexistent content"],
        "starter_packs_skeleton": ["Space"],
    })

    result.finish()
    return result


if __name__ == "__main__":
    result = run()
    result.print_summary()
    sys.exit(result.exit_code)

File: scripts/scenarios/scenarios/20_unspecced_search.ts
/**
 * @module scenarios/20_unspecced_search
 *
 * Scenario: Unspecced Search & Discovery
 *
 * Behavior:
 * - Initialize test accounts and create various posts.
 * - Create a list and a starter pack as search targets.
 * - Perform unspecced search operations (searchActorsSkeleton, searchPostsSkeleton, searchStarterPacksSkeleton).
 *
 * Expectations:
 * - Search operations return successfully for valid and invalid queries.
 * - Search results correctly include actors, posts, and starter packs created in the scenario.
 */

import { ScenarioResult, timedCall } from "@garazyk/hamownia";
export { ScenarioResult, StepResult, StepStatus } from "@garazyk/hamownia";
export type { ScenarioReport } from "@garazyk/hamownia";
import { assert } from "@garazyk/hamownia";
import { XrpcClient, XrpcError } from "@garazyk/gruszka";
import type { ScenarioContext } from "@garazyk/hamownia/config";

function now() {
  return new Date().toISOString();
}

/**
 * Executes the scenario logic.
 * @returns A promise that resolves to the scenario result
 */
export async function run(ctx: ScenarioContext): Promise<ScenarioResult> {
  const result = new ScenarioResult("Unspecced Search & Discovery");
  result.start();

  const client = new XrpcClient(ctx.pds1);

  await timedCall(result, "Server health check", async () => {
    await client.waitForHealthy(30);
  });

  if (result.failed > 0) {
    result.finish();
    return result;
  }

  const charNames = ["luna", "marcus", "rosa"];
  for (const name of charNames) {
    const char = ctx.getCharacter(name);
    const session = await timedCall(
      result,
      `Create account: ${char.name}`,
      async () => {
        return await client.accounts.createAccount(
          char.handle,
          char.email,
          char.password,
        );
      },
      (s) => `did=${s.did}`,
    );
    if (session) {
      char.did = session.did;
      char.accessJwt = session.accessJwt;
    }
  }

  const active = charNames.filter((n) => ctx.getCharacter(n).did);
  for (const name of active) {
    const char = ctx.getCharacter(name);
    try {
      await client.records.createRecord(
        char.did,
        "app.bsky.actor.profile",
        {
          $type: "app.bsky.actor.profile",
          displayName: char.name,
          description: char.persona,
        },
        char.accessJwt,
      );
    } catch (e) {
      if (!(e instanceof XrpcError && e.status === 404)) throw e;
    }
  }

  const luna = ctx.getCharacter("luna");
  const marcus = ctx.getCharacter("marcus");
  const rosa = ctx.getCharacter("rosa");

  const postData = [
    {
      name: "luna",
      texts: [
        "The Orion Nebula is absolutely stunning tonight! #astrophotography",
        "Just published my deep space photography guide #astronomy",
      ],
    },
    {
      name: "marcus",
      texts: [
        "ATProto is the future of decentralized social networking",
        "Building a firehose consumer in Go — streaming thousands of events per second",
      ],
    },
    {
      name: "rosa",
      texts: [
        "Homemade sourdough with roasted garlic and herbs #baking",
        "The best cacio e pepe recipe you will ever try #cooking",
      ],
    },
  ];

  for (const group of postData) {
    const char = ctx.getCharacter(group.name);
    if (char.did && char.accessJwt) {
      for (const text of group.texts) {
        await timedCall(
          result,
          `${char.name} posts`,
          async () => {
            return await client.records.createRecord(
              char.did,
              "app.bsky.feed.post",
              { $type: "app.bsky.feed.post", text, createdAt: now() },
              char.accessJwt,
            );
          },
          (r) => `uri=${r.uri}`,
        );
      }
    }
  }

  if (rosa.did && luna.did && marcus.did) {
    const listRec = await timedCall(
      result,
      "Rosa creates list for starter pack",
      async () => {
        return await client.records.createRecord(
          rosa.did,
          "app.bsky.graph.list",
          {
            $type: "app.bsky.graph.list",
            name: "Space & Code Enthusiasts",
            purpose: "app.bsky.graph.defs#curatelist",
            createdAt: now(),
          },
          rosa.accessJwt,
        );
      },
      (r) => `uri=${r.uri}`,
    );

    if (listRec) {
      await timedCall(
        result,
        "Rosa creates starter pack for search",
        async () => {
          return await client.records.createRecord(
            rosa.did,
            "app.bsky.graph.starterpack",
            {
              $type: "app.bsky.graph.starterpack",
              name: "Space & Code Enthusiasts",
              description: "Friends who love space and technology",
              list: listRec.uri,
              createdAt: now(),
            },
            rosa.accessJwt,
          );
        },
        (r) => `uri=${r.uri}`,
      );
    }
  }

  await new Promise((r) => setTimeout(r, 2000));

  if (marcus.accessJwt) {
    try {
      for (const query of ["nebula", "Luna"]) {
        await timedCall(
          result,
          `Search actors skeleton '${query}'`,
          async () => {
            return await client.search.searchActorsSkeleton(query, {
              token: marcus.accessJwt,
            });
          },
          (r) => `found=${r.actors?.length || 0}`,
        );
      }

      for (const query of ["nebula", "sourdough", "zzz nonexistent content"]) {
        await timedCall(
          result,
          `Search posts skeleton '${query}'`,
          async () => {
            return await client.search.searchPostsSkeleton(query, {
              token: marcus.accessJwt,
            });
          },
          (r) => `found=${r.posts?.length || 0}`,
        );
      }

      await timedCall(
        result,
        "Search starter packs skeleton 'Space'",
        async () => {
          return await client.search.searchStarterPacksSkeleton("Space", {
            token: marcus.accessJwt,
          });
        },
        (r) => `found=${r.starterPacks?.length || 0}`,
      );
    } catch (e) {
      if (!(e instanceof XrpcError && e.status === 404)) throw e;
    }
  }

  result.finish();
  return result;
}

if (import.meta.main) {
  run(createScenarioContext()).then((res) => {
    console.log(res.summary());
    Deno.exit(res.ok ? 0 : 1);
  });
}
